Request Scheduler Prioritization

lilleyse opened the pull request:
Continuation of #3371 for #3241 

I added support for prioritizing requests based on distance. While we want something more generic in the future, I would like some feedback on the approach I'm taking.

The request scheduler uses information about the previous frame to determine the number of requests allowed for each server during the given frame. `resetBudgets` will sort by distance any requests that couldn't be made last frame and estimate new budgets. When a request is made, it can give a yes/no answer by either returning a promise or undefined, as it did before. The request can be sent immediately and not defer until the end of the frame. There may be cases where the view changes drastically between frames and the estimates from the previous frame are off, but this may not be a problem. If it is, I can introduce the concept of stealing from budgets like in `JobScheduler`.

I had to modify a lot of functions in order to send the distance with each request. I considered generating the distance using the x, y, level, and tiling scheme instead, which would make the code cleaner but has its own limitations.

lilleyse commented on Specs/spec-main.js:
```diff
@@ -58,6 +58,8 @@
 
         /*global jasmineRequire,jasmine,exports,specs*/
 
+        Cesium.RequestScheduler.prioritize = false;
```
A lot of the specs fail if prioritization is enabled because it lags behind a frame. So I added this here for simplicity sake, but it's probably a bad place...

pjcozzi replied:
Looks like an OK place to me, I don't think there is a `beforeAll` that applies to all suites.

pjcozzi replied:
Also, add a comment about why this line is here.

pjcozzi commented:
> I considered generating the distance using the x, y, level, and tiling scheme instead, which would make the code cleaner but has its own limitations.

This won't be general enough because, for example, we want to prioritize all requests like glTF assets and their external resources.

pjcozzi commented on Source/Core/ArcGisImageServerTerrainProvider.js:
```diff
@@ -200,11 +202,15 @@ define([
      * @param {Number} x The X coordinate of the tile for which to request geometry.
      * @param {Number} y The Y coordinate of the tile for which to request geometry.
      * @param {Number} level The level of the tile for which to request geometry.
+     * @param {Boolean} [throttleRequests=true] True if the number of simultaneous requests should be limited,
```
When this goes into master - or sooner - include all changes to the public API in CHANGES.md.

pjcozzi commented on Source/Core/ArcGisImageServerTerrainProvider.js:
```diff
@@ -200,11 +202,15 @@ define([
      *          pending and the request will be retried later.
      */
-    ArcGisImageServerTerrainProvider.prototype.requestTileGeometry = function(x, y, level) {
+    ArcGisImageServerTerrainProvider.prototype.requestTileGeometry = function(x, y, level, throttleRequests, distance) {
```
I need to look at the rest of the code, but I wonder if `throttleRequests, distance` should be replaced with something like a single `request` object that doesn't throttle when `undefined`, and has a function to return the `priority` (`distance`).  This would also scale nicely to segmenting 3D Tiles vs. terrain vs. imagery, etc.

At the same time, we don't want to make request attempts too heavily if they are going to be made a lot because of throttling (but there's no concern for the ones that result in an actual request).

pjcozzi replied:
Or maybe to start, just replace both of these with a `requestPriority` that can be `undefined`.

pjcozzi replied:
This would be a breaking change for some functions, but would be easy to deprecate.

pjcozzi replied:
I still think this is a good first step:

> Or maybe to start, just replace both of these with a `requestPriority` that can be `undefined`.

But as I look at more of the code, the long-term approach could be that the `priority` is an object with a bunch of optional properties (`distance` being the most obvious), and the caller fills out as many properties as it knows.  Then the request scheduler knows how to sort based on the incomplete `priority` objects.  This would also allow things like "this request must happen right away (so perhaps we could central all requests even those that happen "immediately").

pjcozzi replied:
Also, leaving this named `distance` is going to make a search and replace later pretty painful.

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -15,8 +15,31 @@ define([
         DeveloperError) {
     "use strict";
 
+    var RequestTypeBudget = function(type) {
```
Why not `function RequestTypeBudget(type) {`?

pjcozzi commented on Source/Core/RequestType.js:
```diff
@@ -0,0 +1,19 @@
+    var RequestType = {
+        TERRAIN : 0,
+        IMAGERY : 1,
+        TILES3D : 2,
```
Down the road we'll need a bit more control here, e.g., what if there are two 3D tilesets: one for a city and one for a point cloud.  Cesium won't know the best priority, but the user might want to override it.

Also, what if a tile has a glTF payload, the glTF requests should be `TILES3D`, but we might also want to bucket them based on geometry vs. texture.

No need to change anything now, but we'll need to think these through.

pjcozzi commented on Source/Scene/Batched3DModel3DTileContentProvider.js:
```diff
@@ -121,7 +124,8 @@ define([
 
-        var promise = RequestScheduler.throttleRequest(this._url, loadArrayBuffer);
+        var distance = this._tile.distanceToCamera;
+        var promise = RequestScheduler.throttleRequest(this._url, loadArrayBuffer, RequestType.TILES3D, distance);
```
As mentioned above, we'll want to thread this through to `Model` whenever it is embedded in a tile.

pjcozzi commented on Source/Scene/GlobeSurfaceTile.js:
```diff
@@ -245,6 +245,13 @@ define([
+                rectangle : tile.rectangle,
+                ellipsoid : tile.tilingScheme.ellipsoid
+            });
+            // Set the distance of the tile now in order to prioritize the request later.
```
Can we also use the min/max height of the parent tile?  I'm afraid this is going to be way off, for example, when zoomed in close on Mount Everest.

@kring any thoughts here?

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -32,6 +55,70 @@ define([
     }
 
+    function handleLeftoverRequest(url, type, distance) {
+        // TODO : don't create a new object, reuse instead
```
Address this and the other `TODO`s before we merge.

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -15,8 +15,31 @@ define([
         DeveloperError) {
```
The new code in this file needs unit tests.

pjcozzi commented:
It would be good to test this with something like the DC buildings and Bing imagery, and see how it compares to before 3D Tiles requests were prioritized.  http://cesiumjs.org/WashingtonDC/

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -32,6 +55,70 @@ define([
+            return;
+        }
+
+        printStats();
```
Remove this.  Instead, for example, the Sandcastle example could call this each frame (or it could be commented out in the Sandcastle example).

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -140,16 +230,33 @@ define([
+        if (RequestScheduler.prioritize && defined(requestType)) {
+            var budget = budgets[server];
+            if (!defined(budget)) {
+                budget = new RequestTypeBudget(requestType);
```
Couldn't two request types go to the same server?

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -32,6 +55,70 @@ define([
+            }
+        }
+
+        // Sort all requests made this frame by distance
```
all?  Or all leftover?

pjcozzi commented:
That's all my comments.

lilleyse commented:
Updated.

I created a `Request` class that is like the `priority` object mentioned in these comments, but also stores additional information about the request and is used inside `RequestScheduler`.

I added the ability to defer a request, so if there aren't any open slots on the current frame it will still return a promise that will be resolved once an opening appears. This is useful for all the requests that aren't imagery/terrain/3d-tiles related. I went through the codebase looking for loading functions like `loadArrayBuffer`, `loadJson`, etc and changed them to use the `RequestScheduler`. @mramato suggested the possibility of doing the inverse approach, and having `loadWithXhr` call `RequestScheduler`.

I haven't tested with the Washington DC buildings yet because I'm having some problems converting the tileset, but I hope to see it working soon. Also `CHANGES.md` still needs to be updated, but I'm waiting until this is final.

pjcozzi commented:
@lilleyse can you merge `3d-tiles` into this?

pjcozzi commented:
> @mramato suggested the possibility of doing the inverse approach, and having loadWithXhr call RequestScheduler.

This was my gut too, but I'm open to other approaches.  Will have a look now.

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,64 @@
+         */
+        this.distance = defaultValue(options.distance, 0.0);
+
+        // Helper members for RequestScheduler
```
Mark these as `@private` since they are being used like C#'s `internal`.

pjcozzi commented on Source/Core/ArcGisImageServerTerrainProvider.js:
```diff
@@ -231,7 +237,17 @@ define([
-        var promise = RequestScheduler.throttleRequest(url, loadImage);
+        if (!defined(request)) {
+            // If a request object isn't provided, perform an immediate request
+            request = new Request();
```
Throughout, for code blocks like these, perhaps:
```javascript
request = new Request({
    defer : true
});
```

lilleyse replied:
The only reason I didn't do this in a few places was to avoid the memory allocation, but maybe it doesn't matter.

pjcozzi replied:
I'm not concerned about it, but leave it if you are.

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -246,7 +250,7 @@ define([
 
         function requestMetadata() {
-            var metadata = loadJson(metadataUrl);
+            var metadata = RequestScheduler.request(metadataUrl, loadJson);
```
Looking at this now, I suspect this is a better approach than doing this implicitly/explicitly in the utility functions.

It's explicit to the user, should extend well to new request functions, and I imagine is what a typical game engine's "resource manager" looks like.

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -484,7 +487,7 @@ define([
      *            returns true.
      */
-    CesiumTerrainProvider.prototype.requestTileGeometry = function(x, y, level, throttleRequests) {
+    CesiumTerrainProvider.prototype.requestTileGeometry = function(x, y, level, request) {
```
This needs some backwards compatibility code for when "`request`" is still a boolean.  We'll also deprecate this - over one release is fine.

pjcozzi commented:
I know this was already merged into `3d-tiles`, but the implementation of `RequestScheduler.hasAvailableRequests` is too slow and allocates memory.  This is an important function that is called on each tile during tree traversal.  Perhaps we can implement some caching (that the caller could store) to avoid the memory allocation and maybe even implicit hash lookup.  I dunno, it might not be perfectly straightforward since, for example, a 3D tileset can reference tiles on many different servers.  Perhaps there is a fast approximation or coarse-grained check we can use.

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -129,41 +256,93 @@ define([
+     * @returns {Promise.<Object>} A Promise for the requested data.
+     */
+    RequestScheduler.request = function(url, requestFunction, parameters) {
+        return RequestScheduler.throttleRequest(new Request({
```
Perhaps this should have a `type` of "OTHER", "USER", "CUSTOM", or something like that, which could be useful for setting priorities and avoiding starvation.

lilleyse commented:
One simple way would be for each content provider to store its server.

pjcozzi commented on Source/Core/EllipsoidTerrainProvider.js:
```diff
@@ -160,14 +160,13 @@ define([
      *          pending and the request will be retried later.
      */
-    EllipsoidTerrainProvider.prototype.requestTileGeometry = function(x, y, level, throttleRequests) {
+    EllipsoidTerrainProvider.prototype.requestTileGeometry = function(x, y, level, request) {
```
Same deprecation comment throughout.

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,64 @@
+     *
+     * @exports Request
+     *
+     * @private
```
Do we plan to keep `Request` and `RequestScheduler` private when they go into master?

Although their most important use for most apps will be deep inside of Cesium for things that stream, I think they will be useful as a public API for any app that wants to play nice with requests.

Perhaps the best approach is to keep them private when this is first merged into master, and then make them public when `3d-tiles` is merged in case we want to change the interface while working on 3D Tiles.  If you agree, update the 3D Tiles roadmap to add a note to make these public.

lilleyse replied:
I agree with keeping it private for now.

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,64 @@
+        /**
+         * Type of request. Used for more fine-grained priority sorting.
+         */
+        this.requestType = options.requestType;
```
This should probably be `type`, e.g., `request.type` instead of `request.requestType`.

Same comment for `requestFunction`, except we would have to use `func` (I think) so maybe just stay with `requestFunction`.

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -129,41 +256,93 @@ define([
      *
      */
-    RequestScheduler.throttleRequest = function(url, requestFunction) {
+    RequestScheduler.throttleRequest = function(request) {
```
I know this was the existing name, but the most common function should have the most common name.  Perhaps this should be `request`, and the current `request` should be `deferRequest`.

I'm also not 100% sold on the "defer" language throughout, perhaps it maps well to promises, in which case it is fine; otherwise, I feel like `requestThisFrame` and `requestAnyFrame` (or something like that) is better.

pjcozzi replied:
I'm putting too much thought into this, but perhaps better semantics are `schedule()` (today's `throttleRequest`) and `request()` (today's `request`).

lilleyse replied:
Those sound good to me.

pjcozzi commented on Source/Scene/Model.js:
```diff
@@ -1010,7 +1014,7 @@ define([
             gltfCache[cacheKey] = cachedGltf;
 
-            loadArrayBuffer(url, options.headers).then(function(arrayBuffer) {
+            RequestScheduler.request(url, loadArrayBuffer, options.headers).then(function(arrayBuffer) {
```
I think I mentioned this in a past review, but `Model` may be making requests on behalf of 3D Tiles so I think we should allow an override for the request type via a `@private` options parameter on the constructor.

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,64 @@
+/*global define*/
+define([
+        'Core/defaultValue'
```
I think this should be `./defaultValue`.

The tests pass, but I get this when I run Viewer:

![image](https://cloud.githubusercontent.com/assets/782098/12332534/ff17d136-babc-11e5-8723-053993c2e797.png)

pjcozzi commented:
> One simple way would be for each content provider to store its server.

Perhaps each tile would store a reference to a small object that is a new `RequestsAvailable` (or something like that) class.  At initialization, they would ask `RequestScheduler` for this object based on their url, then they can check it instead of the slow path currently in `RequestScheduler`.  Under-the-hood, `RequestScheduler` could do something similar.

The idea is that if we know a request won't go through this frame, we want to learn that as quickly as possible and bail on tree traversal.  Likewise, we want requests that may go through to be queued up as quickly as possible with minimal scheduler overhead.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -317,14 +327,17 @@ define([
+        }
     }
 
     function hasAvailableRequests(tiles3D) {
```
I believe this line is wrong:
```javascript
return RequestScheduler.hasAvailableRequests(tiles3D._url)
```
This assumes that the tile's content has the same base path as tileset.json, right?  They could be on two separate servers.

lilleyse commented:
Updated.

pjcozzi commented on Source/Core/ArcGisImageServerTerrainProvider.js:
```diff
@@ -231,7 +237,19 @@ define([
         }
 
-        var promise = RequestScheduler.throttleRequest(url, loadImage);
+        // TODO : request used to be a boolean called throttleRequest. Continue to handle that case until it is deprecated.
```
Remove this `TODO` and friends, and submit an issue for the deprecation.

pjcozzi commented on CHANGES.md:
```diff
@@ -10,6 +10,8 @@ Change Log
 * Breaking changes
     * Removed `OpenStreetMapImageryProvider`. Use `createOpenStreetMapImageryProvider` instead.
+* Deprecated
+    * Deprecated `throttleRequests` flag in `TerrainProvider`. It is replaced by an optional `Request` object that stores information used to prioritize requests.
```
What version will it be removed in?  1.19 is find IMO.

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,80 @@
+        /**
+         * The URL to request.
+         *
+         * @private
```
Since the class is private, each member does not need to be `private`.  I would just mark ones private if they are only intended to be used by `RequestScheduler` (looks like `startPromise` and `server`).

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -5,18 +5,77 @@ define([
+        this.serverName = serverName;
+    }
+
+    RequestServer.prototype.hasAvailableRequests = function() {
```
Do you handle starvation here as we discussed offline?

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -57,15 +170,34 @@ define([
+    RequestScheduler.getRequestServer = function(url) {
+        var requestServer;
+        var serverName = RequestScheduler.getServerName(url);
+        var length = activeRequestsByServer.length;
```
Why is `activeRequestsByServer` an array now?  Wouldn't it be better as an object used as a hashmap?

lilleyse replied:
Yeah, an object makes more sense. I changed it back.

pjcozzi commented on Source/Scene/Cesium3DTile.js:
```diff
@@ -179,6 +183,7 @@ define([
             this.hasContent = false;
         }
         this._content = content;
+        this._requestServer = requestServer;
```
Throughout, access this through a private readonly getter.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -344,22 +354,25 @@ define([
-    function hasAvailableRequests(tiles3D) {
-        return RequestScheduler.hasAvailableRequests(tiles3D._url);
+    function hasAvailableRequests(tile) {
+        return !defined(tile._requestServer) || tile._requestServer.hasAvailableRequests();
```
For my knowledge, when would this not be `defined`?  Data URI?

pjcozzi replied:
Also perhaps this should be a prototype function on `Cesium3DTile`?

lilleyse replied:
It's not defined when the tile is empty, but I suppose also with data URI when we have them.

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,80 @@
+         *
+         * @private
+         */
+        this.server = options.server;
```
This would have more precise semantics if it was named `requestServer` as it is elsewhere.

pjcozzi replied:
Maybe not.  Up to you.

pjcozzi commented:
Just those comments.

Are you happy with how this worked for the DC dataset?

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -344,22 +354,25 @@ define([
+        if (!tile.isContentUnloaded()) {
+            var stats = tiles3D._statistics;
+            ++stats.numberOfPendingRequests;
+            addLoadProgressEvent(tiles3D);
```
This call keeps track of the number of requests so it is possible that it could increase the number of pending requests here even though the scheduler doesn't make the request.

At one point we talked about adding an event (or promise or whatever) that fires when the request is actually made (or not).  Perhaps we should use that to call `addLoadProgressEvent`.

Note that this should all happen before events are raised at the end of the frame with `frameState.afterRender`.

pjcozzi replied:
I would also be fine (perhaps even prefer) raising this event once per frame instead of per tile.  The former is probably better for a UI, and the later for debugging.

lilleyse commented:
I cleaned up the code and have been testing with the DC dataset. When testing with prioritization enabled vs. disabled, there are some small differences but its hard to say that one method is actually better than another. It's easier to notice the benefits in contrived cases where the maximum requests is set to some low number. Usually both 3D Tiles and Imagery/Terrain servers have enough open slots to do their work with or without budgeting.

I also tested the difference for when `Cesium3DTileset` checks if there are available requests vs. always sending requests (slower, but creates accurate budgets). Like above, its hard to see an actual visual difference. I started working on the starvation issue, but in order to do it correctly accounting for distance and the distance ranges of other budgets, it just seemed too complicated for little gain.

lilleyse commented:
This is ready from my end.

pjcozzi commented:
Looks good.  Please cherry pick the commits for a PR into master and update the roadmap as we discussed offline.

lilleyse commented:
Is this ready to be merged here?

pjcozzi commented:
Yes.
